graylog2-web-interface/src/components/rules/RuleForm.jsx

```diff
@@ -34,19 +34,33 @@ const RuleForm = ({ create }) => {
     ruleSourceRef,
   } = useContext(PipelineRulesContext);
 
+  const [isDirty, setIsDirty] = useState(false);
+
   const handleSubmit = (event) => {
     event.preventDefault();
-    handleSavePipelineRule(() => { history.goBack(); });
+
+    handleSavePipelineRule(() => {
+      setIsDirty(false);
+      history.goBack();
+    });
   };
 
   const handleApply = () => {
-    handleSavePipelineRule((rule) => { history.replace(Routes.SYSTEM.PIPELINES.RULE(rule.id)); });
+    handleSavePipelineRule((rule) => {
+      setIsDirty(false);
+      history.replace(Routes.SYSTEM.PIPELINES.RULE(rule.id));
+    });
   };
 
   const handleDescriptionChange = (event) => {
+    setIsDirty(true);
     handleDescription(event.target.value);
   };
 
+  const handleSourceChange = () => {
+    setIsDirty(true);
+  };
+
   const handleCancel = () => {
     history.goBack();
```

Comment:
It seems that this is resetting any changes made in the editor:

https://user-images.githubusercontent.com/716185/108244521-97ea4580-714f-11eb-92f4-b12b360c7457.mp4

Reply:
Oh no! Seems like the whole Pipeline Context is biting our ass once more 😅

Maybe I need to more this dirty flag in there too. 

Good find!